### PR TITLE
CowDexAgSolver solver

### PR DESCRIPTION
**Pull request by josojo**
When I looked into "better ways to merge" last Friday, I realised that actually, its better to do a two step approach:
1. first solve all orders just with paraswap
2. Identify cows and remove the cow-volume from orders
3. Solve aggregated-orders without cowvolume again using zeroEx.

I got hooked and coded up this CowDexAgSovler over the weekend.
It's already producing cool solutions that are settling: here [simulated only on tenderly](https://dashboard.tenderly.co/gp-v2/staging/simulator/48765ec3-feaf-4bca-8c1e-b0527fb2938f):
image

Unfortunately, it's taking too long to finish and clean up the code. Hence, I will keep this PR in draft mode and restart on eco-system friday and focus on affiliate program again.

### Test Plan
```
execute ignored tests
```
and listen to prod:
```
cargo run -p solver -- \                     
  --node-url "https://mainnet.infura.io/v3/$INFURA_PROJECT_ID" \
  --orderbook-url https://protocol-mainnet.gnosis.io \
  --solvers CowDexAgSolver \
  --transaction-strategy DryRun
  ```
  
 PS: This solver has the weakness that if there is a ring of orders, it will just return an empty settlement. Martin has a even better plan for this case :)

### CODE REVIEW

**Comment by fleupold** [comment]
If I remember correct @koeppelmann was working on something similar.

File: solver/src/solver/cow_dex_ag_solver.rs [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+    // todo: Solver does not yet deal with unwrap and warps
+    async fn solve(&self, Auction { orders, .. }: Auction) -> Result<Vec<Settlement>> {
+        if orders.is_empty() {
+            return Ok(Vec::new());
+        }
+        // Randomize which orders we start with to prevent us getting stuck on bad orders.
+        // orders.shuffle(&mut rand::thread_rng())
+        let mut orders: Vec<LimitOrder> = orders.into_iter().collect();
+        // for now, lets only solve for up to 5 orders
+        orders.truncate(5);
```
Comment (nlordell) [inline solver/src/solver/cow_dex_ag_solver.rs:73]: Why limit the amount of orders? Shouldn't we be limiting the amount of pairs instead?

**Comment by nlordell** [comment]
Cool!

**Review by bh2smith — COMMENTED** [review]
Very exciting. Will hold off on review until you feel it is in better shape.

**Comment by josojo** [comment]
I will rewrite it as external http solver

**Pull request closed without merging**